core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2025 LY Corporation
+ *
+ * LY Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry.limiter;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.internal.common.InternalGrpcWebTrailers;
+
+/**
+ * Retry limiter based on a token bucket algorithm with 3 parameters:
+ * <ul>
+ *   <li>Max tokens: Max value of tokens that can be stored in the bucket</li>
+ *   <li>Threshold: The min number tokens in the bucket required to allow retries to happen</li>
+ *   <li>TokenRatio: The number of tokens that a successful attempt adds to the bucket.</li>
+ * </ul>
+ *
+ * <p>The algorithm subtracts 1 token from the bucket for each retriable error and then, if we
+ * can retry (due to timeouts and max attempts), it checks if the bucket is over the threshold.
+ * On every successful request, we add X tokens where X equals to the token ratio.
+ *
+ * <p>In Grpc implementation, the threshold is hardcoded as half the capacity of the bucket but you can set your
+ * own threshold here.
+ *
+ * <p>Internally, the implementation stores all values multiplied by 1000, some manipulation is required as
+ * max tokens can be odd, and when setting the threshold to half of max tokens, rounding will be needed, this
+ * way we keep full precision.
+ *
+ * <p>The implementation is heavily based on GRPC Java <a href="https://github.com/grpc/grpc-java/blob/94532a6b56076c56fb9278e9195bba1190a9260d/core/src/main/java/io/grpc/internal/RetriableStream.java#L1463">implementation</a>.
+ */
+public class GrpcRetryLimiter implements RetryLimiter {
```

Comment:
This implementation tries to mimic gRPC Java as much as possible. There are 2 points that are not implemented. 
gRPC java has support for [pushback](https://github.com/grpc/proposal/blob/4f833c5774e71e94534f72b94ee1b9763ec58516/A6-client-retries.md?plain=1#L201) header, as that is not implemented in the RetryingClient, I have not added support for it here.

The other difference is that gRPC Java considers errors (for reducing tokens) every status defined in the hedging policy as [nonFatalStatusCodes](https://github.com/grpc/proposal/blob/4f833c5774e71e94534f72b94ee1b9763ec58516/A6-client-retries.md?plain=1#L193). We don't have hedging implemented, but the same behaviour can be replicated with the extended constructor passing the complete list of statuses.